### ISSUE TITLE
SAarray fails to construct from data when dtype is explicitly specified

### ISSUE THREAD

**Issue by Jarvi-Izana** [apple/turicreate#2088]
The test failed when passing a map type.

```
        # map input type
        toy_data = SFrame({'a': range(100)})
        map_result = map(lambda x: x+1, [1, 30])
        result = toy_data.filter_by(map_result, 'a')
>       self.assertEqual(len(result), 2)
E       AssertionError: 1 != 2
```

minimal repro:
```
map_result = map(lambda x: x+1, [1, 30])
g = tc.SArray(map(lambda x: x+1, [1, 30]), int)
TypeError                                 Traceback (most recent call last)
<ipython-input-40-04fb1819e4fc> in <module>
----> 1 g = tc.SArray(map(lambda x: x+1, [1, 30]), int)

~/Work/guicreate/debug/src/python/turicreate/data_structures/sarray.py in __init__(self, data, dtype, ignore_cast_failure, _proxy)
    424                     self.__proxy__.load_from_iterable(data, dtype, ignore_cast_failure)
    425             else:
--> 426                 raise TypeError("Unexpected data source. " \
    427                                 "Possible data source types are: list, " \
    428                                 "numpy.ndarray, pandas.Series, and string(url)")

TypeError: Unexpected data source. Possible data source types are: list, numpy.ndarray, pandas.Series, and string(url)
```

**Comment by TobyRoseman** [comment]
@Jarvi-Izana - I'm assuming this is a Python 3 issues. Can you confirm what version of Python you are using?

**Comment by Jarvi-Izana** [comment]
passed the internal test.

**Comment by Jarvi-Izana** [comment]
I'm using python 3.6.

**Comment by Jarvi-Izana** [comment]
Solved by #2089
